### PR TITLE
Msp support

### PR DESCRIPTION
**Pull request by weihou**
* call graphql to get card funding eligibility
* check the graphql response on render; throw error if ineligible.

### CODE REVIEW

File: src/component.js [hunk 1]
```diff
@@ -127,83 +135,100 @@ export const HostedFields = {
...
     // toodoo - revert change below when config is being passed correctly
     const configuration = (typeof __hosted_fields__ !== 'undefined') ? __hosted_fields__.serverConfig : TESTING_CONFIGURATION;
     configuration.assetsUrl = TESTING_CONFIGURATION.assetsUrl;
 
-    const cardVendors = (configuration.fundingEligibility && configuration.fundingEligibility.card && configuration.fundingEligibility.card.vendors) || {};
-    const eligibleCards = Object.keys(cardVendors).filter(key => cardVendors[key].eligible);
+    return (getUccEligibility || graphql.getFundingEligibility()).then((eligibilityData) => {
+      if (!eligibilityData || !eligibilityData.card || !eligibilityData.card.eligible || eligibilityData.card.branded) {
+        // inEligible
+        return ZalgoPromise.reject(new Error('hosted fields is not eligible'));
```
Comment (bluepnume) [inline src/component.js]: Are we planning to return a specific error code here for merchants to consume?

File: src/graphql.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
+/* @flow */
+/* eslint import/no-default-export: off */
+
+import { getLogger, getClientID, getBuyerCountry, getCurrency, getMerchantID, getDisableFunding, getDisableCard, getIntent, getCommit, getVault } from '@paypal/sdk-client/src';
+import { ZalgoPromise } from 'zalgo-promise/src';
+import { request } from 'belter/src';
+
+const GRAPHQL_URI = '/graphql';
+
+function callGraphQL<T>({ query, variables = {}, headers = {} } : { query : string, variables? : { [string] : mixed }, headers? : { [string] : string } }) : ZalgoPromise<T> {
```
Comment (bluepnume) [inline src/graphql.js]: Can we put this in sdk-client so it's in a common place?

File: src/graphql.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+type FundingEligibility = {|
+  card? : ?{
+    eligible? : ?boolean,
+    branded? : ?boolean,
+    vendors? : ?VendorEligibility
+  }
+|};
+
+function getFundingEligibility() : ZalgoPromise<FundingEligibility> {
```
Comment (bluepnume) [inline src/graphql.js]: Also this, seems like it would be good to have in sdk-client

File: src/graphql.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+              commit:$commit,
+              vault:$vault,
+              disableFunding:$disableFunding,
+              disableCard:$disableCard,
+              merchantId:$merchantID
+            ) {
+                card{
+                  eligible 
+                  branded 
+                  vendors{
```
Comment (bluepnume) [inline src/graphql.js]: Do we need to re-fetch all this here?

File: src/graphql.js (hunk 3 above)
Comment (bluepnume) [reply src/graphql.js]: Maybe with an option to override the fields that are requested.

File: src/graphql.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        variables
+      }).then((gqlResult) => {
+        if (!gqlResult || !gqlResult.fundingEligibility) {
+          throw new Error(`GraphQL fundingEligibility returned no fundingEligibility object`);
+        }
+        return gqlResult && gqlResult.fundingEligibility;
+      });
+    }
+    catch (err) {
+      getLogger().error(`GraphQL fundingEligibility error`, err);
```
Comment (bluepnume) [inline src/graphql.js]: Error should be in the form of `gql_fundingeligibility_error` so it can be used as a cal event name

File: src/graphql.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      return ZalgoPromise.reject(err);
+    }
+  }
+  
+  // return default
+  return ZalgoPromise.resolve(__hosted_fields__.serverConfig.fundingEligibility);
```
Comment (bluepnume) [inline src/graphql.js]: Can you use `__FUNDING_ELIGIBILITY__` -- trying to move away from `__hosted_fields__` module-specific approach.

File: test/component.js [hunk 7]
```diff
@@ -51,6 +74,8 @@ describe('hosted-fields-component', () => {
     fakeHostedFieldsInstance = td.object([ 'tokenize' ]);
     td.when(fakeHostedFieldsInstance.tokenize(td.matchers.isA(Object))).thenResolve(fakeTokenizationPayload);
     hostedFieldsCreate = td.replace(hostedFields, 'create');
+    graphqlGetFundingEligibility = td.replace(graphql, 'getFundingEligibility');
```
Comment (bluepnume) [inline test/component.js]: Do we plan a specific test for msp case?

File: src/graphql.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+  mastercard? : ?CardVendorEligibility,
+  amex : ?CardVendorEligibility,
+  discover? : ?CardVendorEligibility,
+  hiper? : ?CardVendorEligibility,
+  elo? : ?CardVendorEligibility,
+  jcb? : ?CardVendorEligibility,
+  cup? : ?CardVendorEligibility
+|};
+
+type FundingEligibility = {|
```
Comment (bluepnume) [inline src/graphql.js]: All of these types can be imported from sdk-client already I think

File: npm-debug.log [hunk 9]
```diff
@@ -0,0 +1,47 @@
+0 info it worked if it ends with ok
```
Comment (bluepnume) [inline npm-debug.log]: let's gitignore this file.

**Comment by bluepnume** [comment]
Few comments, everything else looks good to me. Thanks!

File: package.json [hunk 10]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@paypal/card-components",
-  "version": "1.0.37",
+  "version": "1.0.38-alpha.0",
```
Comment (bluepnume) [inline package.json]: Can you remove this prior to merge

Comment (weihou) [reply package.json]: fixed

**Pull request merged**